File: huggingplace-python-sdk/huggingplace_sdk/huggingplace.py
"""
Main HuggingPlace SDK class
"""

import json
import time
import uuid
from typing import Dict, Any, Optional, Callable, Union
import requests
from .validation import validate_config, validate_log_options
from .errors import AuthenticationError, NetworkError, create_error_from_response
from .session import Session


class HuggingPlace:
    """
    Main HuggingPlace SDK class
    """
    
    def __init__(self, config: Dict[str, Any]):
        """
        Create a new HuggingPlace instance.
        
        Args:
            config: Configuration object
                - api_key: API key for authentication
                - org_id: Organization ID
                - base_url: Base URL for API (optional)
                - mode: Environment mode (prod/dev) (optional)
                - timeout: Request timeout in milliseconds (optional)
                - silent: Silent mode (no console logs) (optional)
        """
        validate_config(config)

        self.config = {
            "mode": "prod",
            "timeout": 10000,
            "silent": False,
            **config,
        }

        self.base_url = config.get("base_url") or "https://anvsj57nul.execute-api.ap-south-1.amazonaws.com"

        # Create session with default headers
        self.session = requests.Session()
        self.session.headers.update({
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.config['api_key']}",
        })
        self.session.timeout = self.config["timeout"] / 1000  # Convert to seconds

    async def log(self, options: Dict[str, Any]) -> None:
        """
        Log a complete interaction with HuggingPlace.
        
        Args:
            options: Log options
                - user_prompt: User prompt
                - ai_response: AI response (preferred over 'response')
                - response: AI response (alternative to 'ai_response')
                - user_uuid: User UUID
                - file_name: File name
                - session_id: Session ID
                - llm_model: LLM model used
                - llm_model2: Secondary LLM model
                - token_count: Token count
                - metaData: Additional metadata
                - user_roles: User roles array
                - org_uuid: Organization UUID
                - mapping_table: Database mapping table
                - step_data: Array of processing steps
                - response_time: Response time in milliseconds
                - message_id: Message ID
                - user_meta_data: User metadata (preferred over 'user_metadata')
                - user_metadata: User metadata (alternative to 'user_meta_data')
                - org_id: Organization ID
                - mode: Environment mode
        """
        try:
            validate_log_options(options)

            # Map response to ai_response if not provided (keep both)
            if "response" in options and "ai_response" not in options:
                options["ai_response"] = options["response"]

            # Map user_metadata to user_meta_data if not provided
            if "user_metadata" in options and "user_meta_data" not in options:
                options["user_meta_data"] = options.pop("user_metadata")

            # Send response_time as-is without any formatting

            # Add org_id and mode from config
            payload = {
                **options,
                "org_id": self.config["org_id"],
                "mode": self.config["mode"],
            }

            # if not self.config["silent"]:
            #     print("📤 Sending payload to backend:", json.dumps(payload, indent=2))

            response = self.session.post(
                f"{self.base_url}/v2/chatgpt/store_generated_response",
                json=payload
            )
            
            if not self.config["silent"]:
                print("📥 Response status:", response.status_code)
                print("📥 Response data:", response.json())

            if response.status_code != 200:
                error_data = response.json() if response.content else {}
                message = error_data.get("message") or error_data.get("error") or "Unknown error"
                raise create_error_from_response(response.status_code, message)

            success_message = response.json().get("message", "Successfully logged interaction")
            if not self.config["silent"]:
                print(f"✅ Logged interaction to HuggingPlace: {success_message}")
                
        except requests.exceptions.RequestException as e:
            if not self.config["silent"]:
                print("❌ Failed to log interaction to HuggingPlace:", str(e))
            raise NetworkError(f"Network error: {str(e)}")
        except Exception as e:
            if not self.config["silent"]:
                print("❌ Failed to log interaction to HuggingPlace:", str(e))
            raise

    async def log_step(self, options: Dict[str, Any]) -> None:
        """
        Log individual processing steps.
        
        Args:
            options: Step options
                - step_name: Step name
                - status: Step status
                - time_ms: Time in milliseconds
                - user_question: User question for this step
                - prompt_response: Response for this step
                - llm_model: LLM model used
                - token: Token count
                - response_time: Response time
                - input_tokens: Input token count
                - output_tokens: Output token count
        """
        # For individual steps, we create a minimal log entry
        await self.log({
            "user_prompt": options.get("user_question", ""),
            "ai_response": options.get("prompt_response", ""),
            "llm_model": options.get("llm_model"),
            "token_count": options.get("token"),
            "response_time": options.get("response_time"),
            "step_data": [options],
        })

    def start_session(
        self, 
        session_id: Optional[str] = None, 
        options: Optional[Dict[str, Any]] = None
    ) -> Session:
        """
        Start a new session for tracking multiple interactions.
        
        Args:
            session_id: Session ID (auto-generated if not provided)
            options: Session options
                - metadata: Default metadata for the session
                - user_metadata: Default user metadata for the session
                
        Returns:
            Session instance
        """
        session_id = session_id or str(uuid.uuid4())
        return Session(self, session_id, options or {})

    async def log_with_timing(
        self, 
        user_prompt: str, 
        response_generator: Callable, 
        options: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Log with automatic timing.
        
        Args:
            user_prompt: User prompt
            response_generator: Function that generates the response
            options: Additional options
            
        Returns:
            The generated response
        """
        start_time = time.time()
        options = options or {}

        try:
            response = await response_generator()
            response_time = time.time() - start_time

            await self.log({
                "user_prompt": user_prompt,
                "ai_response": response,
                "response_time": f"0 min {response_time:.2f} sec",  # Send as string format
                **options,
            })

            return response
        except Exception as error:
            # Log the error as well
            await self.log({
                "user_prompt": user_prompt,
                "ai_response": f"Error: {str(error)}",
                "response_time": f"0 min {(time.time() - start_time):.2f} sec",  # Send as string format
                "metaData": {
                    **options.get("metaData", {}),
                    "error": True,
                    "error_message": str(error),
                },
                **options,
            })

            raise

    def get_config(self) -> Dict[str, Any]:
        """
        Get current configuration (without sensitive data).
        
        Returns:
            Safe configuration object
        """
        safe_config = self.config.copy()
        safe_config.pop("api_key", None)
        return safe_config

    def update_config(self, new_config: Dict[str, Any]) -> None:
        """
        Update configuration.
        
        Args:
            new_config: New configuration options
        """
        updated_config = {**self.config, **new_config}
        validate_config(updated_config)

        self.config = updated_config

        # Update session with new config
        self.session.headers["Authorization"] = f"Bearer {self.config['api_key']}"
        self.session.timeout = self.config["timeout"] / 1000

    async def test_connection(self) -> bool:
        """
        Test connection to HuggingPlace API.
        
        Returns:
            True if connection successful
        """
        try:
            # Try to make a minimal request to test the connection
            test_payload = {
                "user_prompt": "connection_test",
                "ai_response": "test_response",
                "org_id": self.config["org_id"],
                "mode": self.config["mode"]
            }
            
            response = self.session.post(
                f"{self.base_url}/v2/chatgpt/store_generated_response",
                json=test_payload
            )
            return response.status_code == 200
        except Exception as e:
            if not self.config["silent"]:
                print("Connection test failed:", str(e))
            return False 

File: huggingplace-sdk/src/huggingplace.js
import axios from 'axios';
import { v4 as uuidv4 } from 'uuid';
import { validateConfig, validateLogOptions } from './validation.js';
import { AuthenticationError, NetworkError, createErrorFromResponse } from './errors.js';
import { Session } from './session.js';

/**
 * Main HuggingPlace SDK class
 */
export class HuggingPlace {
  /**
   * Create a new HuggingPlace instance
   * @param {Object} config - Configuration object
   * @param {string} config.apiKey - API key for authentication
   * @param {string} config.orgId - Organization ID
   * @param {string} [config.baseUrl] - Base URL for API
   * @param {string} [config.mode] - Environment mode (prod/dev)
   * @param {number} [config.timeout] - Request timeout in milliseconds
   * @param {boolean} [config.silent] - Silent mode (no console logs)
   */
  constructor(config) {
    validateConfig(config);

    this.config = {
      mode: 'prod',
      timeout: 10000,
      silent: false,
      ...config,
    };

    this.baseUrl = config.baseUrl || 'https://anvsj57nul.execute-api.ap-south-1.amazonaws.com';

    this.axiosInstance = axios.create({
      baseURL: this.baseUrl,
      timeout: this.config.timeout,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.config.apiKey}`,
      },
    });

    // Add response interceptor for error handling
    this.axiosInstance.interceptors.response.use(
      (response) => response,
      (error) => {
        if (error.response) {
          const { status, data } = error.response;
          const message = data?.message || data?.error || error.message;
          throw createErrorFromResponse(status, message);
        } else if (error.request) {
          throw new NetworkError();
        } else {
          throw error;
        }
      }
    );
  }

  /**
   * Log a complete interaction with HuggingPlace
   * @param {Object} options - Log options
   * @param {string} options.user_prompt - User prompt
   * @param {string} options.ai_response - AI response (preferred over 'response')
   * @param {string} options.response - AI response (alternative to 'ai_response')
   * @param {string} [options.user_uuid] - User UUID
   * @param {string} [options.file_name] - File name
   * @param {string} [options.session_id] - Session ID
   * @param {string} [options.llm_model] - LLM model used
   * @param {string} [options.llm_model2] - Secondary LLM model
   * @param {number} [options.token_count] - Token count
   * @param {Object} [options.metaData] - Additional metadata
   * @param {Array} [options.user_roles] - User roles array
   * @param {string} [options.org_uuid] - Organization UUID
   * @param {string} [options.mapping_table] - Database mapping table
   * @param {Array} [options.step_data] - Array of processing steps
   * @param {number} [options.response_time] - Response time in milliseconds
   * @param {string} [options.message_id] - Message ID
   * @param {Object} [options.user_meta_data] - User metadata (preferred over 'userMetadata')
   * @param {Object} [options.userMetadata] - User metadata (alternative to 'user_meta_data')
   * @param {number} [options.org_id] - Organization ID
   * @param {string} [options.mode] - Environment mode
   * @returns {Promise<void>}
   */
  async log(options) {
    try {
      validateLogOptions(options);

      // Map response to ai_response if not provided (keep both)
      if (options.response && !options.ai_response) {
        options.ai_response = options.response;
      }

      // Map userMetadata to user_meta_data if not provided
      if (options.userMetadata && !options.user_meta_data) {
        options.user_meta_data = options.userMetadata;
      }

      // Send response_time as-is without any formatting

      // Add org_id and mode from config
      const payload = {
        ...options,
        org_id: this.config.orgId,
        mode: this.config.mode,
      };

      // if (!this.config.silent) {
      //   console.log("📤 Sending payload to backend:", JSON.stringify(payload, null, 2));
      // }

      const response = await this.axiosInstance.post(
        '/v2/chatgpt/store_generated_response',
        payload
      );

      if (!this.config.silent) {
        console.log("📥 Response status:", response.status);
        console.log("📥 Response data:", response.data);
      }

      if (response.status !== 200) {
        throw createErrorFromResponse(response.status, response.data?.message || 'Unknown error');
      }

      const successMessage = response.data.message || 'Successfully logged interaction';
      if (!this.config.silent) {
        console.log(`✅ Logged interaction to HuggingPlace: ${successMessage}`);
      }
    } catch (error) {
      if (!this.config.silent) {
        console.error('❌ Failed to log interaction to HuggingPlace:', error);
      }
      throw error;
    }
  }

  /**
   * Log individual processing steps
   * @param {Object} options - Step options
   * @param {string} options.step_name - Step name
   * @param {string} options.status - Step status
   * @param {number} options.time_ms - Time in milliseconds
   * @param {string} options.userQuestion - User question for this step
   * @param {string} options.promptResponse - Response for this step
   * @param {string} [options.llmModel] - LLM model used
   * @param {number} [options.token] - Token count
   * @param {number} [options.responseTime] - Response time
   * @param {number} [options.inputTokens] - Input token count
   * @param {number} [options.outputTokens] - Output token count
   * @returns {Promise<void>}
   */
  async logStep(options) {
    // For individual steps, we create a minimal log entry
    await this.log({
      user_prompt: options.userQuestion || '',
      ai_response: options.promptResponse || '',
      llm_model: options.llmModel,
      token_count: options.token,
      response_time: options.responseTime,
      step_data: [options],
    });
  }

  /**
   * Start a new session for tracking multiple interactions
   * @param {string} [sessionId] - Session ID (auto-generated if not provided)
   * @param {Object} [options] - Session options
   * @param {Object} [options.metaData] - Default metadata for the session
   * @param {Object} [options.user_meta_data] - Default user metadata for the session
   * @returns {Session} Session instance
   */
  startSession(sessionId, options = {}) {
    const id = sessionId || uuidv4();
    return new Session(this, id, options);
  }

  /**
   * Log with automatic timing
   * @param {string} userPrompt - User prompt
   * @param {Function} responseGenerator - Function that generates the response
   * @param {Object} [options] - Additional options
   * @returns {Promise<string>} The generated response
   */
  async logWithTiming(userPrompt, responseGenerator, options = {}) {
    const startTime = Date.now();

    try {
      const response = await responseGenerator();
      const responseTime = (Date.now() - startTime) / 1000; // Convert to seconds

      await this.log({
        user_prompt: userPrompt,
        ai_response: response,
        response_time: `0 min ${responseTime.toFixed(2)} sec`, // Send as string format
        ...options,
      });

      return response;
    } catch (error) {
      // Log the error as well
      await this.log({
        user_prompt: userPrompt,
        ai_response: `Error: ${error instanceof Error ? error.message : String(error)}`,
        response_time: `0 min ${((Date.now() - startTime) / 1000).toFixed(2)} sec`, // Send as string format
        metaData: {
          ...options.metaData,
          error: true,
          errorMessage: error instanceof Error ? error.message : String(error),
        },
        ...options,
      });

      throw error;
    }
  }

  /**
   * Get current configuration (without sensitive data)
   * @returns {Object} Safe configuration object
   */
  getConfig() {
    const { apiKey, ...safeConfig } = this.config;
    return safeConfig;
  }

  /**
   * Update configuration
   * @param {Object} newConfig - New configuration options
   */
  updateConfig(newConfig) {
    const updatedConfig = { ...this.config, ...newConfig };
    validateConfig(updatedConfig);

    this.config = updatedConfig;

    // Update axios instance with new config
    this.axiosInstance.defaults.headers['Authorization'] = `Bearer ${this.config.apiKey}`;
    this.axiosInstance.defaults.timeout = this.config.timeout;
  }

  /**
   * Test connection to HuggingPlace API
   * @returns {Promise<boolean>} True if connection successful
   */
  async testConnection() {
    try {
      // Try to make a minimal request to test the connection
      // Since /health doesn't exist, we'll test with a minimal log request
      const testPayload = {
        user_prompt: 'connection_test',
        ai_response: 'test_response',
        org_id: this.config.orgId,
        mode: this.config.mode
      };

      const response = await this.axiosInstance.post(
        '/v2/chatgpt/store_generated_response',
        testPayload
      );
      return response.status === 200;
    } catch (error) {
      if (!this.config.silent) {
        console.error('Connection test failed:', error);
      }
      return false;
    }
  }
} 